main.go

```diff
@@ -13,122 +13,144 @@
 		}
 	}
 	if *flagIgnore != "" {
 		var err error
 		if ignoreRe, err = regexp.Compile(*flagIgnore); err != nil {
-			fmt.Println("bad ignore regexp:", err)
-			os.Exit(1)
+			return fmt.Errorf("bad ignore regexp: %s", err)
 		}
 	}
+
+	c := make(chan os.Signal)
+	defer close(c)
+	signal.Notify(c, os.Interrupt)
+	defer signal.Stop(c)
+	var wg sync.WaitGroup
+	defer wg.Wait()
+	ctx, cancel := func() (context.Context, context.CancelFunc) {
+		ctx := context.Background()
+		if *flagMaxTime > 0 {
+			return context.WithTimeout(ctx, *flagMaxTime)
+		}
+		return context.WithCancel(ctx)
+	}()
+	defer cancel()
+	go func() {
+		for range c {
+			cancel()
+		}
+	}()
+
 	startTime := time.Now()
-	finishTime := time.Now().Add(*flagMaxTime)
+
 	res := make(chan []byte)
-	var exitFlag int32
+	wg.Add(*flagP)
 	for i := 0; i < *flagP; i++ {
-		go func() {
-			for atomic.LoadInt32(&exitFlag) == 0 {
-				cmd := exec.Command(flag.Args()[0], flag.Args()[1:]...)
-				done := make(chan bool)
-				if *flagTimeout > 0 {
-					go func() {
-						select {
-						case <-done:
-							return
-						case <-time.After(*flagTimeout):
-						}
-						if !*flagKill {
-							fmt.Printf("process %v timed out\n", cmd.Process.Pid)
-							return
+		go func(ctx context.Context) {
+			defer wg.Done()
+			for ctx.Err() == nil {
+				res <- func() []byte {
+					var cmd *exec.Cmd
+					if *flagTimeout > 0 {
+						if *flagKill {
+							var cancel context.CancelFunc
+							ctx, cancel = context.WithTimeout(ctx, *flagTimeout)
+							defer cancel()
+						} else {
+							defer time.AfterFunc(*flagTimeout, func() {
+								fmt.Printf("process %v timed out\n", cmd.Process.Pid)
+							}).Stop()
 						}
-						cmd.Process.Signal(syscall.SIGABRT)
-						select {
-						case <-done:
-							return
-						case <-time.After(10 * time.Second):
-						}
-						cmd.Process.Kill()
-					}()
-				}
-				out, err := cmd.CombinedOutput()
-				close(done)
-				if err != nil && (failureRe == nil || failureRe.Match(out)) && (ignoreRe == nil || !ignoreRe.Match(out)) {
-					out = append(out, fmt.Sprintf("\n\nERROR: %v\n", err)...)
-				} else {
-					out = []byte{}
-				}
-				res <- out
+					}
+					cmd = exec.CommandContext(ctx, flags.Args()[0], flags.Args()[1:]...)
+					out, err := cmd.CombinedOutput()
+					if err != nil && (failureRe == nil || failureRe.Match(out)) && (ignoreRe == nil || !ignoreRe.Match(out)) {
+						out = append(out, fmt.Sprintf("\n\nERROR: %v\n", err)...)
+					} else {
+						out = []byte{}
+					}
+					return out
+				}()
 			}
-		}()
+		}(ctx)
 	}
 	runs, fails := 0, 0
 	ticker := time.NewTicker(5 * time.Second).C
-	for atomic.LoadInt32(&exitFlag) == 0 {
+	for {
 		select {
 		case out := <-res:
 			runs++
-			if ((*flagMaxTime > 0) && (time.Now().After(finishTime))) ||
-				((*flagMaxRuns > 0) && (runs >= *flagMaxRuns)) {
-				atomic.StoreInt32(&exitFlag, 1)
+			if *flagMaxRuns > 0 && runs >= *flagMaxRuns {
+				cancel()
```

Comment:
This causes `stress` to erroneously report `FAIL` when `maxruns` runs have completed:

```go
./stress -timeout 10m -maxruns 8 true
context canceled
FAIL
```